## ECUACIONES DE ESTADO
##### EJEMPLO 2: Mezclas


---


Este notebook corresponde a un ejemplo en donde se aplica la resolución de las ecuaciones de estado para obtener diferentes variables relacionadas con la cantidad de materia y diversas propiedades termodinámicas de mezclas.


Es recomendable revisar el archivo teórico (EOS_2_mezclas.ipynb) para entender el funcionamiento de este programa y las formulas que se utilizan para cada ecuación de estado para calcular mezclas. Así como es necesario revisar el archivo "README.md" para entender con mayor detalle acerca en que casos es mejor tomar en cuenta ciertas ecuaciones de estado.


Si se requiere modificar las tablas de resultados finales para cada métodos o en el resumen de resultados generales, vea el archivo "README.md" para ver como modificar los parámetros de salida.


**NOTA IMPORTANTE**: Para ejecutar este programa es necesario tener las librerías numpy y tabulate.



---


### RESOLUCIÓN DEL EJERCICIO


Un cilindro cuyo volumen es de 3 $ft^{3}$ contiene aire a 60°C y 4100 $\frac{lb}{in^{2}}$. Calcular la cantidad de aire que está contenida en el cilindro.


Para el ejercicio se tomará en cuenta que la composición del aire consiste en 78% de nitrógeno molecular ($N_{2}$), 21% de oxígeno molecular ($O_{2}$) y 1% de argón (Ar). Entonces se tomará en cuenta los siguientes datos para la resolución de las ecuaciones de estado.


DATOS DEL EJERCICIO


|Compuesto|Masa molar ($\overline{M},\;\frac{g}{mol}$)|Fracción mol|Presión crítica ($P_{C},\;$atm)|Temperatura crítica ($T_{C},\;$K)|Factor acéntrico ($\omega$)|
|----------|----------|----------|-|-|-|
| $N_{2}$ | 28.013 | 0.78 | 33.4567 | 126.2 | 0.039 |
| $O_{2}$ | 31.999 | 0.21 | 49.7409 | 154.6 | 0.025 |
| $Ar$    | 39.948 | 0.01 | 48.0631 | 150.8 | 0.001 |  


Fuente: Reid, R. C., Prausnitz, J. M., & Poling, B. E. (1987). The properties of gases & liquids (4 th edition). McGraw-Hill.


- Temperatura (T) = 60°C = 333.15 K
- Presión (P) = 4100$\;\frac{lb}{in^{2}}\;$= 278.9885 atm
- Volumen (V) = 3 $ft^{3}\;$ = 84.9505 L

---


### _**INGRESO DE DATOS**_


A continuación se muestra el panel de ingreso de datos para obtener los resultados correspondientes para cada una de las ecuaciones de estado.


Se recomienda no cambiar el valor R (constante de los gases ideales), de hacerlo es necesario buscar valores de los datos a ingresar en las unidades que sean consistentes con los del nuevo R. 


El nombre de la sustancia (variable `COMP`) se escribe entre las comillas simples(`'oxígeno'`,`'ACETONA'`). En el ingreso de datos de los multicomponentes asegurese de tener la misma cantidad de datos que busca calcular y separelos con una coma:


(`['COMPUESTO 1','COMPUESTO 2', 'COMPUESTO 3',...]`)

In [33]:
import numpy as np
import tabulate as tb

R=0.082 #(atm*L)/(mol*K) - CONSTANTE DE LOS GASES IDEALES

T=333.15                           #(K)            - TEMPERATURA DEL SISTEMA DE REFERENCIA
P=278.9885                         #(atm)          - PRESIÓN DEL SISTEMA DE REFERENCIA
V=84.9505                          #(L)            - VOLUMEN DEL SISTEMA DE REFERENCIA

COMP=['nitrógeno mol.','oxígeno mol.','Argón']  #NOMBRE DEL COMPUESTO
MM_IND=[28.013, 31.999, 39.948]     #(g/mol)        - MASA MOLAR DEL COMPUESTO
yi=[0.78, 0.21, 0.01]               #(ADIMENSIONAL) - FRACCIÓN MOL DE CADA COMPONENTE (LA SUMATORIA DEBE DAR 1)
PC=[33.4567, 49.7409, 48.0631]      #(atm)          - PRESIÓN CRITICO PARA EL COMPUESTO
TC=[126.2, 154.6, 150.8]            #(K)            - TEMPERATURA CRITICO PARA EL COMPUESTO
w_fa=[0.039, 0.025, 0.001]          #(ADIMENSIONAL) - FACTOR ACENTRICO

if len(COMP)==len(MM_IND)==len(yi)==len(PC)==len(TC)==len(w_fa):
    if np.sum(yi) ==1:
        MMOLAR_GAS=0
        for i in range(len(yi)):
            MMOLAR_GAS += yi[i] * MM_IND[i]
        print(f"LA MASA MOLAR DE LA MEZCLA DE GAS ES DE %5.3f g/mol\n\n"%(MMOLAR_GAS))
    else:
        print("*** ERROR LA SUMA DE LAS FRACCIONES MOLARES DEBE SER 1 ***\n\n")
    print('*** DATOS INGRESADOS AL PROGRAMA ***')
else:
    print("                  *** ERROR ***")
    print("---------------------------------------------------")
    print("  REVISE QUE LAS VARIALES QUE TENGAN ARREGLOS "'[]'"")
    print("      DEBEN TENER LA MISMA CANTIDAD DE DATOS")


LA MASA MOLAR DE LA MEZCLA DE GAS ES DE 28.969 g/mol


*** DATOS INGRESADOS AL PROGRAMA ***


---


# Ecuación del gas ideal


En el caso de la ecuación de los gases ideales no hace distinción entre los compuestos por lo que las ecuaciones no toman en cuenta las interacciones moleculares dentro de la mezcla de gases. A partir de la ecuación del gas ideal:


\begin{equation}
    P\overline{V}=nRT
\end{equation}


Las variables que se presentan en las ecuaciones:  
$P=$ Presión del sistema de referencia  
$T=$ Temperatura del sistema de referencia  
$\overline{V}=$ Volumen molar del gas  
$R =$ Constante del gas ideal  
$n =$ Cantidad de moles del compuesto


Como primer paso es obtener el volumen molar, a partir de la suposición de que el número de moles $n$ sea 1, entonces el volumen molar queda como:


\begin{equation}
    \overline{V}=\frac{RT}{P}
\end{equation}


Como segundo paso es obtener la cantidad de gas tanto en moles como masa del mismo, sin embargo, considerando las modificaciones de las ecuaciones anteriores entonces el número de moles es 1 y la masa de la mezcla (no confundir con masa molar) es el mismo que del dato del masa molar pero en unidades de masa.


\begin{equation}
    n=\frac{P\overline{V}}{RT}\;=\;1\;mol
\end{equation}


\begin{equation}
    m=(1\;mol)\;\overline{M}
\end{equation}


Entonces por tanto por resultado es que se obtiene que solo hay un mol de sustancia y la cantidad de masa de gas será la misma cantidad que la masa molar del mismo.


La masa molar de la mezcla se puede definir con la siguiente ecuación que muestra sumatoria del producto entre la fracción mol de cada componente con la masa molar de cada especie química.


\begin{equation}
    \overline{M}\;=\;\sum\;\overline{M_{i}}\;y_{i}
\end{equation}


Las variables que se presentan en la ecuación:  
$\overline{M} =$ Masa molar de la mezcla gaseosa  
$\overline{M_{i}} =$ Masa molar de cada especie química  
$y_{i} =$ Fracción mol de cada especie química  


Sin embargo, si se resuelve la ecuación completa solo que la diferencia que en vez del volumen molar ($\overline{V})$ se resuelve con el volumen del cilindro (V) que aparece al principio del apartado se obtiene lo siguientes resultados de moles y masa del gas.

In [34]:
print("*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DEL GAS IDEAL ***\n")
VM_EGI=(R*T)/P
MOL_EGI=(P*V)/(R*T)
MAS_EGI=MMOLAR_GAS*MOL_EGI

# - SALIDA DE RESULTADOS
print("RESULTADOS PARA LA ECUACIÓN DEL GAS IDEAL")
print("--------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB2_RES = ['VOLUMEN MOLAR DEL FLUIDO \n(V, L/mol)','CANTIDAD DE MOLES\nDE SUSTANCIA (mol)','MASA DE SUSTANCIA\n(g)']
R_VAR_EGI = [VM_EGI, MOL_EGI, MAS_EGI]
FIL_R_EGI = []
for i in range(3):
    FIL_R_EGI.append([VAR_TAB2_RES[i],R_VAR_EGI[i]])
print(tb.tabulate(FIL_R_EGI, headers=ENC_RES, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))
print("\n------------------------------------------------------\n\n\n")


*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DEL GAS IDEAL ***

RESULTADOS PARA LA ECUACIÓN DEL GAS IDEAL
--------------------------------------------------------

╒═══════════════════════════╤══════════════╕
│         VARIABLE          │  RESULTADO   │
╞═══════════════════════════╪══════════════╡
│ VOLUMEN MOLAR DEL FLUIDO  │   0.097919   │
│        (V, L/mol)         │              │
├───────────────────────────┼──────────────┤
│     CANTIDAD DE MOLES     │  867.558105  │
│    DE SUSTANCIA (mol)     │              │
├───────────────────────────┼──────────────┤
│     MASA DE SUSTANCIA     │ 25132.646431 │
│            (g)            │              │
╘═══════════════════════════╧══════════════╛

------------------------------------------------------





Para la ecucación del gas ideal se obtiene que en el cilindro hay _**867.55 moles o 25.132 kg de aire**_ contenidos en el mismo, sin embargo, como se mencionó anteriormente este cálculo no toma en cuenta el volumen molar que tenga la mezcla incluyendo las interacciones intemoleculares que tengan entre las especies químicas. Con el primer método descrito que toma en cuenta el volumen molar, en donde la mezcla hay 1 mol y la masa del gas será de 28.969 g, sin embargo, este cálculo no es real ni proporcional a los datos que plantean el problema.

---


# Ecuación de estado de Van der Waals


A continuación se muestra los pasos para realizar los cálculos correspondientes para la ecuación de estado de Van der Waals para resolver el ejercicio. Los resultados de cada paso se muestran en los datos de salida de la celda del código.


1. Se obtienen los parámetros de la ecuación de estado $a$ y $b$ para cada compuesto


\begin{equation}
    a\;=\;\frac{27R^{2}T^{2}_{C}}{64P_{C}}\;\;\;\;\;\;\;\;b\;=\;\frac{RT_{C}}{8P_{C}}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:   
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado  
$b =$ Parámetro b para la ecuación de estado  
$T_{C}=$ Temperatura crítica de cada sustancia  
$P_{C}=$ Presión crítica de cada sustancia


2. Se obtienen los parámetros globales $a$ y $b$ para mezcla mediante regla de mezclado que se muestra en continuación.


En el caso del parámetro $a$ cuando se resuelve la expresión cuadrática (izquierda) se generan dos casos debido a que cuando hay interacciones de la misma (centro) ó diferente especie química (derecha) tienen diferentes comportamientos a nivel molecular tanto de atracción o de repulsión que tengan las moléculas.


\begin{equation}
    a\;=\;\left(\sum\;y_{i}\sqrt{a_{i}}\right)^{2}\;=\;\sum\;y_{i}^{2}a_{i}\;=\;\sum\;\sum\;y_{i}\;y_{j}\;\sqrt{a_{i}\;a_{j}}
\end{equation}


- $a =$ Parámetro a global para la ecuación de estado  
- $a_{i}\;ó\;a_{j} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i}\;ó\;y_{j} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


En este caso en el ejercicio se tiene 3 especies químicos entonces se debe se resuelve de la siguiente manera el parámetro global $a$.


\begin{equation}
    a\;=\;y_{1}^{2}\;a_{1}\;+\;y_{1}\;y_{2}\;\sqrt{a_{1}\;a_{2}}\;+\;y_{1}\;y_{3}\;\sqrt{a_{1}\;a_{3}}\;+\;y_{2}\;y_{1}\;\sqrt{a_{2}\;a_{1}}\;+\;y_{2}^{2}\;a_{2}\;+\;y_{2}\;y_{3}\;\sqrt{a_{2}\;a_{3}}\;+\;y_{3}\;y_{1}\;\sqrt{a_{3}\;a_{1}}\;+\;y_{3}\;y_{2}\;\sqrt{a_{3}\;a_{2}}\;+\;y_{3}^{2}\;a_{3}
\end{equation}


Simplificando la expresión teniendo en cuenta que los cuando los subíndices son diferentes (expresión para especies químicas diferentes) se aplica la propiedad conmutativa.


\begin{equation}
    a\;=\;y_{1}^{2}\;a_{1}\;+2\;\left(y_{1}\;y_{2}\;\sqrt{a_{1}\;a_{2}}\right)\;+\;2\;\left(y_{1}\;y_{3}\;\sqrt{a_{1}\;a_{3}}\right)\;+\;y_{2}^{2}\;a_{2}\;+\;2\;\left(y_{2}\;y_{3}\;\sqrt{a_{2}\;a_{3}}\right)\;+\;y_{3}^{2}\;a_{3}
\end{equation}


En el caso del parámetro $b$ de la ecuación de estado, se muestra la siguiente expresión.


\begin{equation}
    b\;=\;\sum\;b_{i}\;y_{i}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
- $b =$ Parámetro b global para la ecuación de estado
- $b_{i} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


Este se resuelve de manera simple, ya que solo es la sumatoria de cada producto entre la fracción mol y el valor del parámetro $b$ individual de cada compuesto. Como se muestra a continuación en el contexto de las condiciones del ejercicio.


\begin{equation}
    b\;=\;b_{1}\;y_{1}\;+\;b_{2}\;y_{2}\;+\;b_{3}\;y_{3}
\end{equation}


3. Una vez obtenido los parámetros globales de la ecuación de estado se resuelve la ecuación polinomial en términos del volumen molar ($\overline{V}$) sustituyendo todos los datos tanto de presión y temperatura del sistema; constante de los gases ideales; y los parámetros $a$ y $b$.


\begin{equation}
    \overline{V^{3}}\;-\;\left(\;b\;+\;\frac{RT}{P}\;\right)\;\overline{V^{2}}\;+\;\left(\;\frac{a}{P}\;\right)\;\overline{V}\;-\;\left(\;\frac{ab}{P}\;\right)\;=\;0
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
$P=$ Presión del sistema de referencia  
$T=$ Temperatura del sistema de referencia  
$\overline{V}=$ Volumen molar del gas  
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado (en este caso el global para la mezcla)  
$b =$ Parámetro b para la ecuación de estado (en este caso el global para la mezcla)


4. Una vez obtenido los coeficientes de ecuación polinomial solo es obtener la raíz real única del misma. (NOTA: Hay en ciertos casos en donde se puede obtener tres raíces reales solo se toma la mayor que corresponde el volumen molar del gas.)


5. Luego de haber obtenido el volumen molar ($\overline{V}$) se puede obtener cantidad de moles ($n$) que hay en el cilindro y por tanto la cantidad de masa del gas ($m$) a partir de las siguientes ecuaciones con sus respectivos despejes.


\begin{equation}
    \overline{V}\;=\;\frac{V}{n}\;\;\;\;\;\;\;\;n\;=\;\frac{V}{\overline{V}}
\end{equation}


\begin{equation}
    \overline{M}\;=\;\frac{m}{n}\;\;\;\;\;\;\;\;m\;=\;n\;\overline{M}
\end{equation}


La masa molar de la mezcla se puede definir con la siguiente ecuación que muestra sumatoria del producto entre la fracción mol de cada componente con la masa molar de cada especie química.


\begin{equation}
    \overline{M}\;=\;\sum\;\overline{M_{i}}\;y_{i}
\end{equation}


Las variables que se presentan en la ecuación:  
$\overline{M} =$ Masa molar de la mezcla gaseosa  
$\overline{M_{i}} =$ Masa molar de cada especie química  
$y_{i} =$ Fracción mol de cada especie química  

In [35]:
print("*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE VAN DER WAALS ***\n\n")

# - CALCULAR LAS CONSTANTE a Y b PARA COMPONENTE DE LA MEZCLA
a_VW=[]
b_VW=[]

for i in range(len(COMP)):
    a_VW.append((27*pow(R,2)*pow(TC[i],2))/(64*PC[i]))
    b_VW.append((R*TC[i])/(8*PC[i]))

print("1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE \nINDIVIDUAL PARA LA ECUACIÓN DE ESTADO DE Van Der Waals")
print("---------------------------------------------------------------------\n")
ENC_VW= ["NC", "COMPONENTE", "CONSTANTE a", "CONSTANTE b"]
FIL_VW= []
for i in range(len(COMP)):
    FIL_VW.append([i+1, COMP[i].upper(), a_VW[i], b_VW[i]])
print(tb.tabulate(FIL_VW, headers=ENC_VW, tablefmt="fancy_grid", floatfmt="3.6f", stralign="center", numalign="center"))
print("\n\n")

# - CALCULAR CONSTANTE a y b GLOBAL MEDIANTE UNA REGLA DE MEZCLADO
a_GLOBAL_VW=0
for i in range(len(a_VW)):
    for j in range(len(a_VW)):
        if i != j:
            a_GLOBAL_VW += np.sqrt(a_VW[i] * a_VW[j]) * yi[i] * yi[j]
        else:
            a_GLOBAL_VW += a_VW[i] * yi[i]**2

b_GLOBAL_VW=0
for i in range(len(b_VW)):
    b_GLOBAL_VW += yi[i]*b_VW[i]

print(f"2. OBTENCIÓN DE LAS CONSTANTES a y b GLOBALES PARA LA MEZCLA \nPARA LA ECUACIÓN DE ESTADO DE Van der Waals")
print("---------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB1_RES = ['CONSTANTE a','CONSTANTE b']
R_VAR_EEVW = [a_GLOBAL_VW, b_GLOBAL_VW]
FIL_R_EEVW = []
for i in range(2):
    FIL_R_EEVW.append([VAR_TAB1_RES[i],R_VAR_EEVW[i]])
print(tb.tabulate(FIL_R_EEVW, headers=ENC_RES, tablefmt="fancy_grid", floatfmt="5.6f", stralign="center", numalign="center"))
print('\n\n')

# -  CALCULAR LOS COEFICIENTES DE LA ECUACIÓN DE ESTADO DE VAN DER WAALS EN SU FORMA POLINOMIAL
print("3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO")
print("--------------------------------------------------------------------------------\n")
coeff1_VW=1                                   #TERMINO CÚBICO
coeff2_VW=-1*(b_GLOBAL_VW+((R*T)/P))          #TERMINO CUADRÁTICO
coeff3_VW=(a_GLOBAL_VW/P)                     #TERMINO LINEAL
coeff4_VW=-1*((a_GLOBAL_VW*b_GLOBAL_VW)/P)    #TERMINO INDEPENDIENTE

POL_VW=[coeff1_VW,coeff2_VW,coeff3_VW,coeff4_VW]
print("--> LA ECUACIÓN POLINÓMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:\n")
print("%5.1f V^3 + \t%5.10f V^2 + \t%5.10f V + \t%5.10f = 0\n\n"%(POL_VW[0],POL_VW[1],POL_VW[2],POL_VW[3]))

# - OBTENER LA RAÍZ REAL DE LA ECUACIÓN POLINOMIAL Y POR TANTO EL VOLUMEN MOLAR
EPOL_VW=np.roots(POL_VW)
print("--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:\n")
print(EPOL_VW)
print("\n\n--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:\n")
# - SELECCIÓN DE LA RAIZ REAL DE LA ECUACIÓN POLINOMIAL
RR_VW=sum(np.isreal(EPOL_VW))
if RR_VW == 1:
    VM_VW=EPOL_VW[np.isreal(EPOL_VW)].real[0]
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_VW))
    print("\n\n")
elif RR_VW == 3:
    VM_VW=max(EPOL_VW)
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_VW))
    print("\n\n")
else:
    print("*** ERROR NO HAY SOLUCION PARA ESTA EOS EN ESTAS CONDICIONES, NO HAY RAIZ REAL UNICA ***\n\n")

# - CALCULO DE LOS MOLES Y MASA DE LA MEZCLA
MOL_VW=V/VM_VW
MAS_VW=MMOLAR_GAS*MOL_VW

# - SALIDA DE RESULTADOS
print("5. RESULTADOS PARA LA ECUACIÓN DE ESTADO DE VAN DER WAALS")
print("---------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB2_RES = ['VOLUMEN MOLAR DEL FLUIDO \n(V, L/mol)','CANTIDAD DE MOLES\nDE SUSTANCIA (mol)','MASA DE SUSTANCIA\n(g)']
R_VAR_EEVW = [VM_VW, MOL_VW, MAS_VW]
FIL_R_EEVW = []
for i in range(3):
    FIL_R_EEVW.append([VAR_TAB2_RES[i],R_VAR_EEVW[i]])
print(tb.tabulate(FIL_R_EEVW, headers=ENC_RES, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))
print("\n-------------------------------------------------------------------\n\n\n")


*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE VAN DER WAALS ***


1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE 
INDIVIDUAL PARA LA ECUACIÓN DE ESTADO DE Van Der Waals
---------------------------------------------------------------------

╒══════╤════════════════╤═══════════════╤═══════════════╕
│  NC  │   COMPONENTE   │  CONSTANTE a  │  CONSTANTE b  │
╞══════╪════════════════╪═══════════════╪═══════════════╡
│  1   │ NITRÓGENO MOL. │   1.350352    │   0.038663    │
├──────┼────────────────┼───────────────┼───────────────┤
│  2   │  OXÍGENO MOL.  │   1.363066    │   0.031858    │
├──────┼────────────────┼───────────────┼───────────────┤
│  3   │     ARGÓN      │   1.342154    │   0.032160    │
╘══════╧════════════════╧═══════════════╧═══════════════╛



2. OBTENCIÓN DE LAS CONSTANTES a y b GLOBALES PARA LA MEZCLA 
PARA LA ECUACIÓN DE ESTADO DE Van der Waals
---------------------------------------------------------------------

╒═════════════╤═════════

#### Para la ecuación de estado de Van der Waals se obtiene que en el cilindro hay _**806.82 moles o 23.373 kg de aire**_ en el cilindro.

---


# Ecuación de estado de Redlich-Kwong


A continuación se muestra los pasos para realizar los cálculos correspondientes para la ecuación de estado de Redlich-Kwong para resolver el ejercicio. Los resultados de cada paso se muestran en los datos de salida de la celda del código.


1. Se obtienen los parámetros de la ecuación de estado $a$ y $b$ para cada compuesto


\begin{equation}
    a\;=\;\frac{0.42748\;R^{2}T^{2.5}_{C}}{P_{C}}\;\;\;\;\;\;\;\;b\;=\;\frac{0.08664\;RT_{C}}{P_{C}}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:   
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado  
$b =$ Parámetro b para la ecuación de estado  
$T_{C}=$ Temperatura crítica de cada sustancia  
$P_{C}=$ Presión crítica de cada sustancia


2. Se obtienen los parámetros globales $a$ y $b$ para mezcla mediante regla de mezclado que son las mismas que se presentó en la ecuación de estado de Van der Waals que se muestra en continuación.


En el caso del parámetro $a$ cuando se resuelve la expresión cuadrática (izquierda) se generan dos casos debido a que cuando hay interacciones de la misma (centro) ó diferente especie química (derecha) tienen diferentes comportamientos a nivel molecular tanto de atracción o de repulsión que tengan las moléculas.


\begin{equation}
    a\;=\;\left(\sum\;y_{i}\sqrt{a_{i}}\right)^{2}\;=\;\sum\;y_{i}^{2}a_{i}\;=\;\sum\;\sum\;y_{i}\;y_{j}\;\sqrt{a_{i}\;a_{j}}
\end{equation}


- $a =$ Parámetro a global para la ecuación de estado  
- $a_{i}\;ó\;a_{j} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i}\;ó\;y_{j} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


En este caso en el ejercicio se tiene 3 especies químicos entonces se debe se resuelve de la siguiente manera el parámetro global $a$.


\begin{equation}
    a\;=\;y_{1}^{2}\;a_{1}\;+\;y_{1}\;y_{2}\;\sqrt{a_{1}\;a_{2}}\;+\;y_{1}\;y_{3}\;\sqrt{a_{1}\;a_{3}}\;+\;y_{2}\;y_{1}\;\sqrt{a_{2}\;a_{1}}\;+\;y_{2}^{2}\;a_{2}\;+\;y_{2}\;y_{3}\;\sqrt{a_{2}\;a_{3}}\;+\;y_{3}\;y_{1}\;\sqrt{a_{3}\;a_{1}}\;+\;y_{3}\;y_{2}\;\sqrt{a_{3}\;a_{2}}\;+\;y_{3}^{2}\;a_{3}
\end{equation}


Simplificando la expresión teniendo en cuenta que los cuando los subíndices son diferentes (expresión para especies químicas diferentes) se aplica la propiedad conmutativa.


\begin{equation}
    a\;=\;y_{1}^{2}\;a_{1}\;+2\;\left(y_{1}\;y_{2}\;\sqrt{a_{1}\;a_{2}}\right)\;+\;2\;\left(y_{1}\;y_{3}\;\sqrt{a_{1}\;a_{3}}\right)\;+\;y_{2}^{2}\;a_{2}\;+\;2\;\left(y_{2}\;y_{3}\;\sqrt{a_{2}\;a_{3}}\right)\;+\;y_{3}^{2}\;a_{3}
\end{equation}


En el caso del parámetro $b$ de la ecuación de estado, se muestra la siguiente expresión.


\begin{equation}
    b\;=\;\sum\;b_{i}\;y_{i}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
- $b =$ Parámetro b global para la ecuación de estado
- $b_{i} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


Este se resuelve de manera simple, ya que solo es la sumatoria de cada producto entre la fracción mol y el valor del parámetro $b$ individual de cada compuesto. Como se muestra a continuación en el contexto de las condiciones del ejercicio.


\begin{equation}
    b\;=\;b_{1}\;y_{1}\;+\;b_{2}\;y_{2}\;+\;b_{3}\;y_{3}
\end{equation}


3. Una vez obtenido los parámetros globales de la ecuación de estado se resuelve la ecuación polinomial en términos del volumen molar ($\overline{V}$) sustituyendo todos los datos tanto de presión y temperatura del sistema; constante de los gases ideales; y los parámetros $a$ y $b$.


\begin{equation}
    \overline{V^{3}}\;-\;\left(\;\frac{RT}{P}\;\right)\;\overline{V^{2}}\;+\;\frac{1}{P}\;\left(\;\frac{a}{\sqrt{T}}\;-\;bRT\;-\;Pb^{2}\;\right)\;\overline{V}\;-\;\left(\;\frac{ab}{P\sqrt{T}}\;\right)\;=\;0
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
$P=$ Presión del sistema de referencia  
$T=$ Temperatura del sistema de referencia  
$\overline{V}=$ Volumen molar del gas  
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado (en este caso el global para la mezcla)  
$b =$ Parámetro b para la ecuación de estado (en este caso el global para la mezcla)


4. Una vez obtenido los coeficientes de ecuación polinomial solo es obtener la raíz real única del misma. (NOTA: Hay en ciertos casos en donde se puede obtener tres raíces reales solo se toma la mayor que corresponde el volumen molar del gas.)


5. Luego de haber obtenido el volumen molar ($\overline{V}$) se puede obtener cantidad de moles ($n$) que hay en el cilindro y por tanto la cantidad de masa del gas ($m$) a partir de las siguientes ecuaciones con sus respectivos despejes.


\begin{equation}
    \overline{V}\;=\;\frac{V}{n}\;\;\;\;\;\;\;\;n\;=\;\frac{V}{\overline{V}}
\end{equation}


\begin{equation}
    \overline{M}\;=\;\frac{m}{n}\;\;\;\;\;\;\;\;m\;=\;n\;\overline{M}
\end{equation}


El masa molar de la mezcla se puede definir con la siguiente ecuación que muestra sumatoria del producto entre la fracción mol de cada componente con la masa molar de cada especie química.


\begin{equation}
    \overline{M}\;=\;\sum\;\overline{M_{i}}\;y_{i}
\end{equation}


Las variables que se presentan en la ecuación:  
$\overline{M} =$ Masa molar de la mezcla gaseosa  
$\overline{M_{i}} =$ Masa molar de cada especie química  
$y_{i} =$ Fracción mol de cada especie química  

In [36]:
print("*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE REDLICH-KWONG ***\n")

# - CALCULAR LAS CONSTANTE a Y b PARA COMPONENTE DE LA MEZCLA
a_RK=[]
b_RK=[]

for i in range(len(COMP)):
    a_RK.append((0.42748*pow(R,2)*pow(TC[i],2.5))/PC[i])
    b_RK.append((0.08664*R*TC[i])/PC[i])

print("1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE \nINDIVIDUAL PARA LA ECUACIÓN DE ESTADO DE Redlich-Kwong")
print("---------------------------------------------------------------------\n")
ENC_RK= ["NC", "COMPONENTE", "CONSTANTE a", "CONSTANTE b"]
FIL_RK= []
for i in range(len(COMP)):
    FIL_RK.append([i+1, COMP[i].upper(), a_RK[i], b_RK[i]])
print(tb.tabulate(FIL_RK, headers=ENC_RK, tablefmt="fancy_grid", floatfmt="3.6f", stralign="center", numalign="center"))
print("\n\n")

# - CALCULAR CONSTANTE a y b GLOBAL MEDIANTE UNA REGLA DE MEZCLADO
a_GLOBAL_RK=0
for i in range(len(a_RK)):
    for j in range(len(a_RK)):
        if i != j:
            a_GLOBAL_RK += np.sqrt(a_RK[i] * a_RK[j]) * yi[i] * yi[j]
        else:
            a_GLOBAL_RK += a_RK[i] * yi[i]**2

b_GLOBAL_RK=0
for i in range(len(b_RK)):
    b_GLOBAL_RK += yi[i]*b_RK[i]

print(f"2. OBTENCIÓN DE LAS CONSTANTES a y b GLOBALES PARA LA MEZCLA \nPARA LA ECUACIÓN DE ESTADO DE Redlich-Kwong")
print("------------------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB1_RES = ['CONSTANTE a','CONSTANTE b']
R_VAR_EERK = [a_GLOBAL_RK, b_GLOBAL_RK]
FIL_R_EERK = []
for i in range(2):
    FIL_R_EERK.append([VAR_TAB1_RES[i],R_VAR_EERK[i]])
print(tb.tabulate(FIL_R_EERK, headers=ENC_RES, tablefmt="fancy_grid", floatfmt="5.6f", stralign="center", numalign="center"))
print('\n\n')

# - CALCULAR LOS COEFICIENTES DE LA ECUACIÓN DE ESTADO DE REDLICH-KWONG EN SU FORMA POLINOMIAL
print("3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO")
print("------------------------------------------------------------------------------\n")
coeff1_RK=1                                                                    #TERMINO CÚBICO
coeff2_RK=-1*((R*T)/P)                                                         #TERMINO CUADRÁTICO
coeff3_RK=(1/P)*((a_GLOBAL_RK/T**0.5)-(b_GLOBAL_RK*R*T)-(P*(b_GLOBAL_RK**2)))  #TERMINO LINEAL
coeff4_RK=-1*((a_GLOBAL_RK*b_GLOBAL_RK)/(P*(T**0.5)))                          #TERMINO INDEPENDIENTE

POL_RK=[coeff1_RK,coeff2_RK,coeff3_RK,coeff4_RK]
print("--> LA ECUACIÓN POLINOMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:\n")
print("%5.1f V^3 + \t%5.10f V^2 + \t%5.10f V + \t%5.10f = 0\n\n"%(POL_RK[0],POL_RK[1],POL_RK[2],POL_RK[3]))

# - OBTENER LA RAÍZ REAL DE LA ECUACIÓN POLINOMIAL Y POR TANTO EL VOLUMEN MOLAR
EPOL_RK=np.roots(POL_RK)
print("--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:\n")
print(EPOL_RK)
print("\n\n--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:\n")
# - SELECCIÓN DE LA RAIZ REAL DE LA ECUACIÓN POLINOMIAL
RR_RK=sum(np.isreal(EPOL_RK))
if RR_RK == 1:
    VM_RK=EPOL_RK[np.isreal(EPOL_RK)].real[0]
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_RK))
    print("\n\n")
elif RR_RK == 3:
    VM_RK=max(EPOL_RK)
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_RK))
    print("\n\n")
else:
    print("*** ERROR NO HAY SOLUCION PARA ESTA EOS EN ESTAS CONDICIONES, NO HAY RAIZ REAL UNICA ***\n\n")

# - CALCULO DE LA DENSIDAD Y DEL FACTOR DE COMPRESIBILIDAD PARA LA SUSTANCIA
MOL_RK=V/VM_RK
MAS_RK=MMOLAR_GAS*MOL_RK

# - RESULTADOS DE SALIDA
print("5. RESULTADOS PARA LA ECUACIÓN DE ESTADO DE REDLICH-KWONG")
print("---------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB2_RES = ['VOLUMEN MOLAR DEL FLUIDO \n(V, L/mol)','CANTIDAD DE MOLES\nDE SUSTANCIA (mol)','MASA DE SUSTANCIA\n(g)']
R_VAR_EERK = [VM_RK, MOL_RK, MAS_RK]
FIL_R_EERK = []
for i in range(3):
    FIL_R_EERK.append([VAR_TAB2_RES[i],R_VAR_EERK[i]])
print(tb.tabulate(FIL_R_EERK, headers=ENC_RES, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))
print("\n-------------------------------------------------------------------\n\n\n")


*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE REDLICH-KWONG ***

1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE 
INDIVIDUAL PARA LA ECUACIÓN DE ESTADO DE Redlich-Kwong
---------------------------------------------------------------------

╒══════╤════════════════╤═══════════════╤═══════════════╕
│  NC  │   COMPONENTE   │  CONSTANTE a  │  CONSTANTE b  │
╞══════╪════════════════╪═══════════════╪═══════════════╡
│  1   │ NITRÓGENO MOL. │   15.371236   │   0.026798    │
├──────┼────────────────┼───────────────┼───────────────┤
│  2   │  OXÍGENO MOL.  │   17.173293   │   0.022081    │
├──────┼────────────────┼───────────────┼───────────────┤
│  3   │     ARGÓN      │   16.700716   │   0.022291    │
╘══════╧════════════════╧═══════════════╧═══════════════╛



2. OBTENCIÓN DE LAS CONSTANTES a y b GLOBALES PARA LA MEZCLA 
PARA LA ECUACIÓN DE ESTADO DE Redlich-Kwong
------------------------------------------------------------------------------

╒═════════════╤═

#### Para la ecuación de estado de Redlich-Kwong se obtiene que en el cilindro hay _**802.17 moles o 23.238 kg de aire**_ en el cilindro.

---


# Ecuación de estado de Soave


A continuación se muestra los pasos para realizar los cálculos correspondientes para la ecuación de estado de Soave para resolver el ejercicio. Los resultados de cada paso se muestran en los datos de salida de la celda del código.


1. Se obtienen los parámetros de la ecuación de estado $a$, $b$ y $\alpha$ para cada compuesto


\begin{equation}
    a\;=\;\frac{0.42747\;R^{2}T^{2}_{C}}{P_{C}}\;\;\;\;\;\;\;\;b\;=\;\frac{0.08664\;RT_{C}}{P_{C}}
\end{equation}


\begin{equation}
    \alpha\;=\;\left[1\;+\;\left(\left(0.48508\;+\;1.55171\;\omega\;-\;0.15613\;\omega^{2}\right)\left(1\;-\;T_{r}^{0.5}\right)\right)\right]^{2}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:   
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado  
$b =$ Parámetro b para la ecuación de estado  
$\alpha =$ Parámetro $\alpha$ para la ecuación de estado  
$T_{C}=$ Temperatura crítica de cada sustancia  
$P_{C}=$ Presión crítica de cada sustancia  
$\omega=$ Factor acéntrico de cada sustancia  
$T_{r}\;=\;\dfrac{T}{T_{C}}\;=$ Temperatura reducida es la división entre la temperatura del sistema de referencia y la temperatura crítica de cada componente  


2. Se obtienen los parámetros globales $a\alpha$ y $b$ para mezcla mediante regla de mezclado que son diferentes tomando en cuenta que se requiere un parámetro de interacción binaria $k_{ij}$ (solo es válido para especies químicas diferentes y no pares de compuestos químicos iguales dando como valor predeterminado 0).


En el caso del parámetro $a\alpha$ cuando se resuelve para cada uno de los subíndices correspondientes a la cantidad de componentes de la mezcla se generan dos casos debido a que cuando hay interacciones de la misma (inferior) ó diferente especie química (superior) tienen diferentes comportamientos a nivel molecular tanto de atracción o de repulsión que tengan las moléculas.


\begin{equation}
        a\alpha\;=\;\sum\sum\;y_{i}\;y_{j}\;\left(1\;-\;k_{ij}\right)\;\sqrt{\left(a\alpha\right)_{i}\left(a\alpha\right)_{j}}
\end{equation}


\begin{equation}
        a\alpha\;=\;\sum\;y_{i}^{2}\;\left(1\;-\;0\right)\;\left(a\alpha\right)_{i}
\end{equation}


- $a\alpha =$ Parámetro $a\alpha$ global para la ecuación de estado  
- $(a\alpha)_{i}\;ó\;(a\alpha)_{j} =$ Parámetro producto $a\alpha$ individual para cada compuesto  
- $y_{i}\;ó\;y_{j} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $k_{ij} =$ Parámetro de interacción binaria entre dos compuestos en específico o dos tipos de especies químicas en específicos, solo es válido para especies químicas diferentes y no pares de compuestos químicos iguales dando como valor predeterminado 0. (NOTA: Es probable que en la literatura haya parámetros de interacción binaria para una ecuación de estado en específico ya sea este o cualquier otro incluso la Peng-Robinson. También incluso un parámetro de interacción general para todos los ecuaciones de estado)  


Para este ejercicio los parámetros de interación binaria ($k_{ij}$) son los siguientes:  
- Entre $N_{2}\;y\;O_{2} =$ -0.014
- Entre $N_{2}\;y\;Ar =$ -0.008
- Entre $O_{2}\;y\;Ar =$ 0.021


Fuente: Chapoy, A., Nazeri, M., Kapateh, M., Burgass, R., Coquelet, C., & Tohidi, B. (2013). Effect of impurities on thermophysical properties and phase behaviour of a CO2-rich system in CCS. International Journal Of Greenhouse Gas Control, 19, 92-100. https://doi.org/10.1016/j.ijggc.2013.08.019


En este caso en el ejercicio se tiene 3 especies químicos entonces se debe se resuelve de la siguiente manera el parámetro global $a\alpha$.


\begin{equation}
    a\alpha\;=\;y_{1}^{2}\;\left(a\alpha\right)_{1}\;+\;y_{1}\;y_{2}\;\left(1-k_{1\:2}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{2}}\;+\;y_{1}\;y_{3}\;\left(1-k_{1\:3}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{3}}\;+\;y_{2}\;y_{1}\;\left(1-k_{2\:1}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{1}}\;+\;y_{2}^{2}\;\left(a\alpha\right)_{2}\;
\end{equation}


\begin{equation}
   +\;y_{2}\;y_{3}\;\left(1-k_{2\:3}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{3}}\;+\;y_{3}\;y_{1}\;\left(1-k_{3\:1}\right)\sqrt{\left(a\alpha\right)_{3}\left(a\alpha\right)_{1}}\;+\;y_{3}\;y_{2}\;\left(1-k_{3\:2}\right)\sqrt{\left(a\alpha\right)_{3}\left(a\alpha\right)_{2}}\;+\;y_{3}^{2}\;\left(a\alpha\right)_{3} 
\end{equation}


Simplificando la expresión teniendo en cuenta que los cuando los subíndices son diferentes (expresión para especies químicas diferentes) se aplica la propiedad conmutativa (incluyendo el parámetro de interacción binaria $k_{ij}=k_{ji}$).


\begin{equation}
    a\alpha\;=\;y_{1}^{2}\;\left(a\alpha\right)_{1}\;+\;2\left(\;y_{1}\;y_{2}\;\left(1-k_{1\:2}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{2}}\right)\;+\;2\left(y_{1}\;y_{3}\;\left(1-k_{1\:3}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{3}}\right)\;+\;y_{2}^{2}\;\left(a\alpha\right)_{2}\;+\;2\left(y_{2}\;y_{3}\;\left(1-k_{2\:3}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{3}}\;\right)\;+\;y_{3}^{2}\;\left(a\alpha\right)_{3} 
\end{equation}


En el caso del parámetro $b$ de la ecuación de estado, se muestra la siguiente expresión.


\begin{equation}
    b\;=\;\sum\;b_{i}\;y_{i}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
- $b =$ Parámetro b global para la ecuación de estado
- $b_{i} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


Este se resuelve de manera simple, ya que solo es la sumatoria de cada producto entre la fracción mol y el valor del parámetro $b$ individual de cada compuesto. Como se muestra a continuación en el contexto de las condiciones del ejercicio.


\begin{equation}
    b\;=\;b_{1}\;y_{1}\;+\;b_{2}\;y_{2}\;+\;b_{3}\;y_{3}
\end{equation}


3. Una vez obtenido los parámetros globales de la ecuación de estado se resuelve la ecuación polinomial en términos del volumen molar ($\overline{V}$) sustituyendo todos los datos tanto de presión y temperatura del sistema; constante de los gases ideales; y los parámetros $a$ y $b$.


\begin{equation}
    \overline{V^{3}}\;-\;\left(\;\frac{RT}{P}\;\right)\;\overline{V^{2}}\;+\;\frac{1}{P}\;\left(\;a\;\alpha\;-\;bRT\;-\;Pb^{2}\;\right)\;\overline{V}\;-\;\left(\;\frac{a\alpha b}{P}\;\right)\;=\;0
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
$P=$ Presión del sistema de referencia  
$T=$ Temperatura del sistema de referencia  
$\overline{V}=$ Volumen molar del gas  
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado (en este caso el global para la mezcla)  
$b =$ Parámetro b para la ecuación de estado (en este caso el global para la mezcla)


4. Una vez obtenido los coeficientes de ecuación polinomial solo es obtener la raíz real única del misma. (NOTA: Hay en ciertos casos en donde se puede obtener tres raíces reales solo se toma la mayor que corresponde el volumen molar del gas.)


5. Luego de haber obtenido el volumen molar ($\overline{V}$) se puede obtener cantidad de moles ($n$) que hay en el cilindro y por tanto la cantidad de masa del gas ($m$) a partir de las siguientes ecuaciones con sus respectivos despejes.


\begin{equation}
    \overline{V}\;=\;\frac{V}{n}\;\;\;\;\;\;\;\;n\;=\;\frac{V}{\overline{V}}
\end{equation}


\begin{equation}
    \overline{M}\;=\;\frac{m}{n}\;\;\;\;\;\;\;\;m\;=\;n\;\overline{M}
\end{equation}


El masa molar de la mezcla se puede definir con la siguiente ecuación que muestra sumatoria del producto entre la fracción mol de cada componente con la masa molar de cada especie química.


\begin{equation}
    \overline{M}\;=\;\sum\;\overline{M_{i}}\;y_{i}
\end{equation}


Las variables que se presentan en la ecuación:  
$\overline{M} =$ Masa molar de la mezcla gaseosa  
$\overline{M_{i}} =$ Masa molar de cada especie química  
$y_{i} =$ Fracción mol de cada especie química  

In [37]:
print("*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE SOAVE ***\n")

# - CALCULAR LAS CONSTANTE a, b y alpha PARA COMPONENTE DE LA MEZCLA
TR_S=[]
a_S=[]
b_S=[]
alpha_S=[]

for i in range(len(COMP)):
    TR_S.append(T/TC[i])
    a_S.append((0.42747*pow(R,2)*pow(TC[i],2))/PC[i])
    b_S.append((0.08664*R*TC[i])/PC[i])
    alpha_S.append((1 + ((0.48508 + (1.55171 * w_fa[i]) - (0.15613 * w_fa[i]**2)) * (1 - (TR_S[i]**0.5))))**2)

print("1. OBTENCIÓN DE LAS CONSTANTES a, b y α PARA CADA COMPONENTE INDIVIDUAL PARA LA \nECUACIÓN DE ESTADO DE Soave")
print("------------------------------------------------------------------------------------------\n")
ENC_S= ["NC", "COMPONENTE", "CONSTANTE a", "CONSTANTE b", "CONSTANTE α","TEMPERATURA\nREDUCIDA"]
FIL_S= []
for i in range(len(COMP)):
    FIL_S.append([i+1, COMP[i].upper(), a_S[i], b_S[i], alpha_S[i], TR_S[i]])
print(tb.tabulate(FIL_S, headers=ENC_S, tablefmt="fancy_grid", floatfmt="3.6f", stralign="center", numalign="center"))

# - CALCULAR CONSTANTE a y b GLOBAL MEDIANTE UNA REGLA DE MEZCLADO
print("\n\nINGRESE LOS VALORES DE PARÁMETRO DE INTERACCIÓN BINARIA VALIDO PARA ESTE ECUACIÓN DE ESTADO:")
print("------------------------------------------------------------------------------------------------\n")
n_S= len(a_S)
kij_S= np.zeros((n_S, n_S))

for i in range(n_S):
    for j in range(i, n_S):
        if i == j:
            kij_S[i][j] = 0
        else:
            kij_S[i][j] = float(input(f"ENTRE {COMP[i]} Y {COMP[j]}: "))
            kij_S[j][i] = kij_S[i][j]

a_GLOBAL_S=0
for i in range(len(a_S)):
    for j in range(len(a_S)):
        if i != j:
            a_GLOBAL_S += ((1 - (kij_S[i][j])) * np.sqrt(a_S[i] * alpha_S[i] * a_S[j] * alpha_S[j])) * yi[i] * yi[j]
        else:
            a_GLOBAL_S += a_S[i] * alpha_S[i] * yi[i]**2

b_GLOBAL_S=0
for i in range(len(b_S)):
    b_GLOBAL_S += yi[i]*b_S[i]

print(f"\n\n2. OBTENCIÓN DE LAS CONSTANTES aα y b GLOBALES PARA LA MEZCLA PARA LA \nECUACIÓN DE ESTADO DE Soave")
print("------------------------------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB1_RES = ['CONSTANTE aα GLOBAL','CONSTANTE b GLOBAL']
R_VAR_EES = [a_GLOBAL_S, b_GLOBAL_S]
FIL_R_EES = []
for i in range(2):
    FIL_R_EES.append([VAR_TAB1_RES[i],R_VAR_EES[i]])
print(tb.tabulate(FIL_R_EES, headers=ENC_RES, tablefmt="fancy_grid", floatfmt="5.6f", stralign="center", numalign="center"))
print('\n\n')

# - CALCULAR LOS COEFICIENTES DE LA ECUACIÓN DE ESTADO DE SOAVE EN SU FORMA POLINOMIAL
print("3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO")
print("------------------------------------------------------------------------------------------\n")
coeff1_S=1                                                          #TERMINO CÚBICO
coeff2_S=-1*((R*T)/P)                                               #TERMINO CUADRÁTICO
coeff3_S=(1/P)*((a_GLOBAL_S)-(b_GLOBAL_S*R*T)-(P*(b_GLOBAL_S**2)))  #TERMINO LINEAL
coeff4_S=-1*((a_GLOBAL_S*b_GLOBAL_S)/(P))                           #TERMINO INDEPENDIENTE

POL_S=[coeff1_S,coeff2_S,coeff3_S,coeff4_S]
print("--> LA ECUACIÓN POLINOMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:\n")
print("%5.1f V^3 + \t%5.10f V^2 + \t%5.10f V + \t%5.10f = 0\n\n"%(POL_S[0],POL_S[1],POL_S[2],POL_S[3]))

# - OBTENER LA RAÍZ REAL DE LA ECUACIÓN POLINOMIAL Y POR TANTO EL VOLUMEN MOLAR
EPOL_S=np.roots(POL_S)
print("--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:\n")
print(EPOL_S)
print("\n\n--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:\n")
# - SELECCIÓN DE LA RAIZ REAL DE LA ECUACIÓN POLINOMIAL
RR_S=sum(np.isreal(EPOL_S))
if RR_S == 1:
    VM_S=EPOL_S[np.isreal(EPOL_S)].real[0]
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_S))
    print("\n\n")
elif RR_S == 3:
    VM_S=max(EPOL_S)
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_S))
    print("\n\n")
else:
    print("*** ERROR NO HAY SOLUCION PARA ESTA EOS EN ESTAS CONDICIONES, NO HAY RAIZ REAL UNICA ***\n\n")

# - CALCULO DE LA DENSIDAD Y DEL FACTOR DE COMPRESIBILIDAD PARA LA SUSTANCIA
MOL_S=V/VM_S
MAS_S=MMOLAR_GAS*MOL_S

# - RESULTADOS DE SALIDA
print("5. RESULTADOS PARA LA ECUACIÓN DE ESTADO DE SOAVE")
print("---------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB2_RES = ['VOLUMEN MOLAR DEL FLUIDO \n(V, L/mol)','CANTIDAD DE MOLES\nDE SUSTANCIA (mol)','MASA DE SUSTANCIA\n(g)']
R_VAR_EES = [VM_S, MOL_S, MAS_S]
FIL_R_EES = []
for i in range(3):
    FIL_R_EES.append([VAR_TAB2_RES[i],R_VAR_EES[i]])
print(tb.tabulate(FIL_R_EES, headers=ENC_RES, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))
print("\n-------------------------------------------------------------------\n\n\n")


*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE SOAVE ***

1. OBTENCIÓN DE LAS CONSTANTES a, b y α PARA CADA COMPONENTE INDIVIDUAL PARA LA 
ECUACIÓN DE ESTADO DE Soave
------------------------------------------------------------------------------------------

╒══════╤════════════════╤═══════════════╤═══════════════╤═══════════════╤═══════════════╕
│  NC  │   COMPONENTE   │  CONSTANTE a  │  CONSTANTE b  │  CONSTANTE α  │  TEMPERATURA  │
│      │                │               │               │               │   REDUCIDA    │
╞══════╪════════════════╪═══════════════╪═══════════════╪═══════════════╪═══════════════╡
│  1   │ NITRÓGENO MOL. │   1.368261    │   0.026798    │   0.434649    │   2.639857    │
├──────┼────────────────┼───────────────┼───────────────┼───────────────┼───────────────┤
│  2   │  OXÍGENO MOL.  │   1.381143    │   0.022081    │   0.569863    │   2.154916    │
├──────┼────────────────┼───────────────┼───────────────┼───────────────┼───────────────┤

ENTRE nitrógeno mol. Y oxígeno mol.:  -0.014
ENTRE nitrógeno mol. Y Argón:  -0.008
ENTRE oxígeno mol. Y Argón:  0.021




2. OBTENCIÓN DE LAS CONSTANTES aα y b GLOBALES PARA LA MEZCLA PARA LA 
ECUACIÓN DE ESTADO DE Soave
------------------------------------------------------------------------------------------

╒═════════════════════╤═════════════╕
│      VARIABLE       │  RESULTADO  │
╞═════════════════════╪═════════════╡
│ CONSTANTE aα GLOBAL │  0.637923   │
├─────────────────────┼─────────────┤
│ CONSTANTE b GLOBAL  │  0.025763   │
╘═════════════════════╧═════════════╛



3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO
------------------------------------------------------------------------------------------

--> LA ECUACIÓN POLINOMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:

  1.0 V^3 + 	-0.0979190899 V^2 + 	-0.0008998267 V + 	-0.0000589080 = 0


--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:

[ 0.11083333+0.j         -0.00645712+0.02213158j -0.00645712-0.02213158j]


--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:

VOLUMEN MOLAR DEL FLUIDO (L/mol) = 0.110

#### Para la ecuación de estado de Soave se obtiene que en el cilindro hay _**766.47 moles o 22.204 kg de aire**_ en el cilindro.

---


# Ecuación de estado de Peng-Robinson


A continuación se muestra los pasos para realizar los cálculos correspondientes para la ecuación de estado de Peng-Robinson para resolver el ejercicio. Los resultados de cada paso se muestran en los datos de salida de la celda del código.


1. Se obtienen los parámetros de la ecuación de estado $a$, $b$ y $\alpha$ para cada compuesto


\begin{equation}
    a\;=\;\frac{0.45724\;R^{2}T^{2}_{C}}{P_{C}}\;\;\;\;\;\;\;\;b\;=\;\frac{0.07780\;RT_{C}}{P_{C}}
\end{equation}


\begin{equation}
    \alpha\;=\;\left[1\;+\;\left(\left(0.37464\;+\;1.54226\;\omega\;-\;0.26992\;\omega^{2}\right)\left(1\;-\;T_{r}^{0.5}\right)\right)\right]^{2}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:   
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado  
$b =$ Parámetro b para la ecuación de estado  
$\alpha =$ Parámetro $\alpha$ para la ecuación de estado  
$T_{C}=$ Temperatura crítica de cada sustancia  
$P_{C}=$ Presión crítica de cada sustancia  
$\omega=$ Factor acéntrico de cada sustancia  
$T_{r}\;=\;\dfrac{T}{T_{C}}\;=$ Temperatura reducida es la división entre la temperatura del sistema de referencia y la temperatura crítica de cada componente  


2. Se obtienen los parámetros globales $a\alpha$ y $b$ para mezcla mediante regla de mezclado que son diferentes tomando en cuenta que se requiere un parámetro de interacción binaria $k_{ij}$ (solo es válido para especies químicas diferentes y no pares de compuestos químicos iguales dando como valor predeterminado 0).


En el caso del parámetro $a\alpha$ cuando se resuelve para cada uno de los subíndices correspondientes a la cantidad de componentes de la mezcla se generan dos casos debido a que cuando hay interacciones de la misma (superior) ó diferente especie química (inferior) tienen diferentes comportamientos a nivel molecular tanto de atracción o de repulsión que tengan las moléculas.


\begin{equation}
        a\alpha\;=\;\sum\sum\;y_{i}\;y_{j}\;\left(1\;-\;k_{ij}\right)\;\sqrt{\left(a\alpha\right)_{i}\left(a\alpha\right)_{j}}
\end{equation}


\begin{equation}
        a\alpha\;=\;\sum\;y_{i}^{2}\;\left(1\;-\;0\right)\;\left(a\alpha\right)_{i}
\end{equation}


- $a\alpha =$ Parámetro $a\alpha$ global para la ecuación de estado  
- $(a\alpha)_{i}\;ó\;(a\alpha)_{j} =$ Parámetro producto $a\alpha$ individual para cada compuesto  
- $y_{i}\;ó\;y_{j} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $k_{ij} =$ Parámetro de interacción binaria entre dos compuestos en específico o dos tipos de especies químicas en específicos, solo es válido para especies químicas diferentes y no pares de compuestos químicos iguales dando como valor predeterminado 0. (NOTA: Es probable que en la literatura haya parámetros de interacción binaria para una ecuación de estado en específico ya sea este o cualquier otro incluso la Peng-Robinson. También incluso un parámetro de interacción general para todos los ecuaciones de estado)  


Para este ejercicio los parámetros de interación binaria ($k_{ij}$) son los siguientes:  
- Entre $N_{2}\;y\;O_{2} =$ -0.014
- Entre $N_{2}\;y\;Ar =$ -0.008
- Entre $O_{2}\;y\;Ar =$ 0.021


Fuente: Chapoy, A., Nazeri, M., Kapateh, M., Burgass, R., Coquelet, C., & Tohidi, B. (2013). Effect of impurities on thermophysical properties and phase behaviour of a CO2-rich system in CCS. International Journal Of Greenhouse Gas Control, 19, 92-100. https://doi.org/10.1016/j.ijggc.2013.08.019


En este caso en el ejercicio se tiene 3 especies químicos entonces se debe se resuelve de la siguiente manera el parámetro global $a\alpha$.


\begin{equation}
    a\alpha\;=\;y_{1}^{2}\;\left(a\alpha\right)_{1}\;+\;y_{1}\;y_{2}\;\left(1-k_{1\:2}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{2}}\;+\;y_{1}\;y_{3}\;\left(1-k_{1\:3}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{3}}\;+\;y_{2}\;y_{1}\;\left(1-k_{2\:1}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{1}}\;+\;y_{2}^{2}\;\left(a\alpha\right)_{2}\;
\end{equation}


\begin{equation}
   +\;y_{2}\;y_{3}\;\left(1-k_{2\:3}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{3}}\;+\;y_{3}\;y_{1}\;\left(1-k_{3\:1}\right)\sqrt{\left(a\alpha\right)_{3}\left(a\alpha\right)_{1}}\;+\;y_{3}\;y_{2}\;\left(1-k_{3\:2}\right)\sqrt{\left(a\alpha\right)_{3}\left(a\alpha\right)_{2}}\;+\;y_{3}^{2}\;\left(a\alpha\right)_{3} 
\end{equation}


Simplificando la expresión teniendo en cuenta que los cuando los subíndices son diferentes (expresión para especies químicas diferentes) se aplica la propiedad conmutativa (incluyendo el parámetro de interacción binaria $k_{ij}=k_{ji}$).


\begin{equation}
    a\alpha\;=\;y_{1}^{2}\;\left(a\alpha\right)_{1}\;+\;2\left(\;y_{1}\;y_{2}\;\left(1-k_{1\:2}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{2}}\right)\;+\;2\left(y_{1}\;y_{3}\;\left(1-k_{1\:3}\right)\sqrt{\left(a\alpha\right)_{1}\left(a\alpha\right)_{3}}\right)\;+\;y_{2}^{2}\;\left(a\alpha\right)_{2}\;+\;2\left(y_{2}\;y_{3}\;\left(1-k_{2\:3}\right)\sqrt{\left(a\alpha\right)_{2}\left(a\alpha\right)_{3}}\;\right)\;+\;y_{3}^{2}\;\left(a\alpha\right)_{3} 
\end{equation}


En el caso del parámetro $b$ de la ecuación de estado, se muestra la siguiente expresión.


\begin{equation}
    b\;=\;\sum\;b_{i}\;y_{i}
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
- $b =$ Parámetro b global para la ecuación de estado
- $b_{i} =$ Parámetro a individual para cada compuesto ya sea de interacción de igual o diferente especie química  
- $y_{i} =$ Fracción mol individual para cada compuesto ya sea de interacción de igual o diferente especie química  


Este se resuelve de manera simple, ya que solo es la sumatoria de cada producto entre la fracción mol y el valor del parámetro $b$ individual de cada compuesto. Como se muestra a continuación en el contexto de las condiciones del ejercicio.


\begin{equation}
    b\;=\;b_{1}\;y_{1}\;+\;b_{2}\;y_{2}\;+\;b_{3}\;y_{3}
\end{equation}


3. Una vez obtenido los parámetros globales de la ecuación de estado se resuelve la ecuación polinomial en términos del volumen molar ($\overline{V}$) sustituyendo todos los datos tanto de presión y temperatura del sistema; constante de los gases ideales; y los parámetros $a$ y $b$.


\begin{equation}
    \overline{V^{3}}\;+\;\left(\;b\;-\;\frac{RT}{P}\;\right)\;\overline{V^{2}}\;+\;\left(\;\frac{a\alpha}{P}\;-\;\frac{2bRT}{P}\;-\;3b^{2}\right)\;\overline{V}\;+\;\left(b^{3}\;+\;\frac{RTb^{2}}{P}\;-\;\frac{a\alpha b}{P}\right)\;=\;0
\end{equation}


Las variables que se presentan en las ecuaciones son las siguientes:  
$P=$ Presión del sistema de referencia  
$T=$ Temperatura del sistema de referencia  
$\overline{V}=$ Volumen molar del gas  
$R =$ Constante del gas ideal  
$a =$ Parámetro a para la ecuación de estado (en este caso el global para la mezcla)  
$b =$ Parámetro b para la ecuación de estado (en este caso el global para la mezcla)


4. Una vez obtenido los coeficientes de ecuación polinomial solo es obtener la raíz real única del misma. (NOTA: Hay en ciertos casos en donde se puede obtener tres raíces reales solo se toma la mayor que corresponde el volumen molar del gas.)


5. Luego de haber obtenido el volumen molar ($\overline{V}$) se puede obtener cantidad de moles ($n$) que hay en el cilindro y por tanto la cantidad de masa del gas ($m$) a partir de las siguientes ecuaciones con sus respectivos despejes.


\begin{equation}
    \overline{V}\;=\;\frac{V}{n}\;\;\;\;\;\;\;\;n\;=\;\frac{V}{\overline{V}}
\end{equation}


\begin{equation}
    \overline{M}\;=\;\frac{m}{n}\;\;\;\;\;\;\;\;m\;=\;n\;\overline{M}
\end{equation}


El masa molar de la mezcla se puede definir con la siguiente ecuación que muestra sumatoria del producto entre la fracción mol de cada componente con la masa molar de cada especie química.


\begin{equation}
    \overline{M}\;=\;\sum\;\overline{M_{i}}\;y_{i}
\end{equation}


Las variables que se presentan en la ecuación:  
$\overline{M} =$ Masa molar de la mezcla gaseosa  
$\overline{M_{i}} =$ Masa molar de cada especie química  
$y_{i} =$ Fracción mol de cada especie química  

In [38]:
print("*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE PENG-ROBINSON ***\n")

# - CALCULAR LAS CONSTANTE a Y b PARA COMPONENTE DE LA MEZCLA
TR_PR=[]
a_PR=[]
b_PR=[]
alpha_PR=[]

for i in range(len(COMP)):
    TR_PR.append(T/TC[i])
    a_PR.append((0.45724*pow(R,2)*pow(TC[i],2))/(PC[i]))
    b_PR.append((0.07780*R*TC[i])/(PC[i]))
    alpha_PR.append((1+((0.37464+(1.54226*(w_fa[i]))-(0.26992*(w_fa[i]**2)))*(1-(TR_PR[i]**0.5))))**2)

print("1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE INDIVIDUAL PARA LA \nECUACIÓN DE ESTADO DE Peng-Robinson")
print("------------------------------------------------------------------------------------------\n")
ENC_PR= ["NC", "COMPONENTE", "CONSTANTE a", "CONSTANTE b", "CONSTANTE α","TEMPERATURA\nREDUCIDA"]
FIL_PR= []
for i in range(len(COMP)):
    FIL_PR.append([i+1, COMP[i].upper(), a_PR[i], b_PR[i], alpha_PR[i], TR_PR[i]])
print(tb.tabulate(FIL_PR, headers=ENC_PR, tablefmt="fancy_grid", floatfmt="3.6f", stralign="center", numalign="center"))

# - CALCULAR CONSTANTE a y b GLOBAL MEDIANTE UNA REGLA DE MEZCLADO
print("\n\nINGRESE LOS VALORES DE PARÁMETRO DE INTERACCIÓN BINARIA VALIDO PARA ESTE ECUACIÓN DE ESTADO:")
print("------------------------------------------------------------------------------------------------\n")
n_PR= len(a_PR)
kij_PR= np.zeros((n_PR, n_PR))

for i in range(n_PR):
    for j in range(i, n_PR):
        if i == j:
            kij_PR[i][j] = 0
        else:
            kij_PR[i][j] = float(input(f"ENTRE {COMP[i]} Y {COMP[j]}: "))
            kij_PR[j][i] = kij_PR[i][j]

a_GLOBAL_PR=0
for i in range(len(a_PR)):
    for j in range(len(a_PR)):
        if i != j:
            a_GLOBAL_PR += ((1 - (kij_PR[i][j])) * np.sqrt(a_PR[i] * alpha_PR[i] * a_PR[j] * alpha_PR[j])) * yi[i] * yi[j]
        else:
            a_GLOBAL_PR += a_PR[i] * alpha_PR[i] * (yi[i]**2)

b_GLOBAL_PR=0
for i in range(len(b_PR)):
    b_GLOBAL_PR += yi[i]*b_PR[i]

print(f"\n\n2. OBTENCIÓN DE LAS CONSTANTES aα y b GLOBALES PARA LA MEZCLA PARA LA \nECUACIÓN DE ESTADO DE Peng-Robinson")
print("------------------------------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB1_RES = ['CONSTANTE aα','CONSTANTE b']
R_VAR_EEPR = [a_GLOBAL_PR, b_GLOBAL_PR]
FIL_R_EEPR = []
for i in range(2):
    FIL_R_EEPR.append([VAR_TAB1_RES[i],R_VAR_EEPR[i]])
print(tb.tabulate(FIL_R_EEPR, headers=ENC_RES, tablefmt="fancy_grid", floatfmt="5.6f", stralign="center", numalign="center"))
print('\n\n')

# - CALCULAR LOS COEFICIENTES DE LA ECUACIÓN DE ESTADO DE PENG-ROBINSON EN SU FORMA POLINOMIAL
print("3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO")
print("------------------------------------------------------------------------------------------\n")
coeff1_PR=1                                                                             #TERMINO CÚBICO
coeff2_PR=((b_GLOBAL_PR)-((R*T)/P))                                                     #TERMINO CUADRÁTICO
coeff3_PR=(((a_GLOBAL_PR)/P)-((2*b_GLOBAL_PR*R*T)/P)-(3*(b_GLOBAL_PR**2)))              #TERMINO LINEAL
coeff4_PR=((b_GLOBAL_PR**3)+((R*T*(b_GLOBAL_PR**2))/P)-((a_GLOBAL_PR*b_GLOBAL_PR)/P))   #TERMINO INDEPENDIENTE

POL_PR=[coeff1_PR,coeff2_PR,coeff3_PR,coeff4_PR]
print("--> LA ECUACIÓN POLINOMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:\n")
print("%5.1f V^3 + \t%5.10f V^2 + \t%5.10f V + \t%5.10f = 0\n\n"%(POL_PR[0],POL_PR[1],POL_PR[2],POL_PR[3]))

# - OBTENER LA RAÍZ REAL DE LA ECUACIÓN POLINOMIAL Y POR TANTO EL VOLUMEN MOLAR
EPOL_PR=np.roots(POL_PR)
print("--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:\n")
print(EPOL_PR)
print("\n\n--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:\n")
# - SELECCIÓN DE LA RAIZ REAL DE LA ECUACIÓN POLINOMIAL
RR_PR=sum(np.isreal(EPOL_PR))
if RR_PR == 1:
    VM_PR=EPOL_PR[np.isreal(EPOL_PR)].real[0]
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_PR))
    print("\n\n")
elif RR_PR == 3:
    VM_PR=max(EPOL_PR)
    print("VOLUMEN MOLAR DEL FLUIDO (L/mol) = %5.6f"%(VM_PR))
    print("\n\n")
else:
    print("*** ERROR NO HAY SOLUCION PARA ESTA EOS EN ESTAS CONDICIONES, NO HAY RAIZ REAL UNICA ***\n\n")

# - OBTENIENDO LOS MOLES Y MASA DEL GAS
MOL_PR=V/VM_PR
MAS_PR=MMOLAR_GAS*MOL_PR

# - RESULTADOS DE SALIDA
print("5. RESULTADOS PARA LA ECUACIÓN DE ESTADO DE PENG-ROBINSON")
print("---------------------------------------------------------------------\n")
ENC_RES = ["VARIABLE", "RESULTADO"]
VAR_TAB2_RES = ['VOLUMEN MOLAR DEL FLUIDO (L/mol)','CANTIDAD DE MOLES\nDE SUSTANCIA (mol)','MASA DE SUSTANCIA\n(g)']
R_VAR_EEPR = [VM_PR, MOL_PR, MAS_PR]
FIL_R_EEPR = []
for i in range(3):
    FIL_R_EEPR.append([VAR_TAB2_RES[i],R_VAR_EEPR[i]])
print(tb.tabulate(FIL_R_EEPR, headers=ENC_RES, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))
print("\n-------------------------------------------------------------------\n\n\n")


*** CALCULANDO LAS VARIABLES DEL FLUIDO POR LA ECUACIÓN DE ESTADO DE PENG-ROBINSON ***

1. OBTENCIÓN DE LAS CONSTANTES a y b PARA CADA COMPONENTE INDIVIDUAL PARA LA 
ECUACIÓN DE ESTADO DE Peng-Robinson
------------------------------------------------------------------------------------------

╒══════╤════════════════╤═══════════════╤═══════════════╤═══════════════╤═══════════════╕
│  NC  │   COMPONENTE   │  CONSTANTE a  │  CONSTANTE b  │  CONSTANTE α  │  TEMPERATURA  │
│      │                │               │               │               │   REDUCIDA    │
╞══════╪════════════════╪═══════════════╪═══════════════╪═══════════════╪═══════════════╡
│  1   │ NITRÓGENO MOL. │   1.463550    │   0.024064    │   0.530882    │   2.639857    │
├──────┼────────────────┼───────────────┼───────────────┼───────────────┼───────────────┤
│  2   │  OXÍGENO MOL.  │   1.477329    │   0.019828    │   0.650794    │   2.154916    │
├──────┼────────────────┼───────────────┼───────────────┼───────────────┼───

ENTRE nitrógeno mol. Y oxígeno mol.:  -0.014
ENTRE nitrógeno mol. Y Argón:  -0.008
ENTRE oxígeno mol. Y Argón:  0.021




2. OBTENCIÓN DE LAS CONSTANTES aα y b GLOBALES PARA LA MEZCLA PARA LA 
ECUACIÓN DE ESTADO DE Peng-Robinson
------------------------------------------------------------------------------------------

╒══════════════╤═════════════╕
│   VARIABLE   │  RESULTADO  │
╞══════════════╪═════════════╡
│ CONSTANTE aα │  0.819946   │
├──────────────┼─────────────┤
│ CONSTANTE b  │  0.023134   │
╘══════════════╧═════════════╛



3. CALCULO DE LOS COEFICIENTES DE LA ECUACIÓN POLINOMIAL Y DE LA RAIZ DEL MISMO
------------------------------------------------------------------------------------------

--> LA ECUACIÓN POLINOMIAL CALCULADA EN TÉRMINOS DE VOLUMEN MOLAR:

  1.0 V^3 + 	-0.0747849462 V^2 + 	-0.0031971172 V + 	-0.0000032049 = 0


--> 4. LAS RAICES OBTENIDAS DE LA ECUACIÓN DE POLINOMIAL SON:

[ 0.10540512 -0.02959271 -0.00102746]


--> DE LO OBTENIDO DE LA ECUACIÓN DE ESTADO:

VOLUMEN MOLAR DEL FLUIDO (L/mol) = 0.105405



5. RESULTADOS PARA LA ECUACIÓN DE ESTADO DE PENG-ROBINSON
------------

#### Para la ecuación de estado de Peng-Robinson se obtiene que en el cilindro hay _**805.94 moles o 23.347 kg de aire**_ en el cilindro.

---


# TABLA COMPARATIVA DE RESULTADOS ENTRE LOS DIFERENTES ECUACIONES DE ESTADO

In [39]:
#RESUMEN DE RESULTADOS DE LO OBTENIDO EN ESTE PROGRAMA
print("RESUMEN DE RESULTADOS DE VOLUMEN MOLAR, MOLES Y MASA DEL GAS PURO POR CADA ECUACION DE ESTADO")
print("------------------------------------------------------------------------------------------------\n")

ENC_RVD = ["ECUACIÓN DE\nESTADO", "VOLUMEN MOLAR\n(L/mol)","CANTIDAD DE MOLES\nDE SUSTANCIA (mol)","MASA DE SUSTANCIA\n(g)"]
EOS=['Gas ideal','Van der Waals', 'Redlich-Kwong', 'Soave', 'Peng-Robinson']
VM_SIST=[VM_EGI, VM_VW, VM_RK, VM_S, VM_PR]
MOL_SIST=[MOL_EGI, MOL_VW, MOL_RK, MOL_S, MOL_PR]
MAS_SIST=[MAS_EGI, MAS_VW, MAS_RK, MAS_S, MAS_PR]
FIL_RVD = []
for i in range(5):
    FIL_RVD.append([EOS[i], VM_SIST[i], MOL_SIST[i], MAS_SIST[i]])
print(tb.tabulate(FIL_RVD, headers=ENC_RVD, tablefmt="fancy_grid", floatfmt=".6f", stralign="center", numalign="center"))

RESUMEN DE RESULTADOS DE VOLUMEN MOLAR, MOLES Y MASA DEL GAS PURO POR CADA ECUACION DE ESTADO
------------------------------------------------------------------------------------------------

╒═══════════════╤═════════════════╤══════════════════════╤═════════════════════╕
│  ECUACIÓN DE  │  VOLUMEN MOLAR  │  CANTIDAD DE MOLES   │  MASA DE SUSTANCIA  │
│    ESTADO     │     (L/mol)     │  DE SUSTANCIA (mol)  │         (g)         │
╞═══════════════╪═════════════════╪══════════════════════╪═════════════════════╡
│   Gas ideal   │    0.097919     │      867.558105      │    25132.646431     │
├───────────────┼─────────────────┼──────────────────────┼─────────────────────┤
│ Van der Waals │    0.105290     │      806.826221      │    23373.279596     │
├───────────────┼─────────────────┼──────────────────────┼─────────────────────┤
│ Redlich-Kwong │    0.105900     │      802.176205      │    23238.571362     │
├───────────────┼─────────────────┼──────────────────────┼─────────────────────

Como se muestra en la tabla de resultados comparativa se muestra como varía la cantidad de gas que contiene el cilindro, lo cual puede generar la pregunta de cuál ecuación de estado es mejor representa en condiciones reales ciertas mezclas gases como por ejemplo pueden ser los hidrocarburos u otros compuestos organicos e inorgánicos en fase gaseosa. 


Ahora puedes experimentar agregando más componentes a la mezcla ya sea en este archivo siguiendo el ejercicio con las composiciones restantes del aire atmosférico (agregando los demás gases de composición), o siguiendo otros ejercicios como el siguiente.


_Una mezcla formada por 66.66% en mol en propileno y el resto de dióxido de azufre que está a 425 K y una presión de 60 atm. Obtenga la cantidad de moles y masa contenidos en tanque de 100 litros._


También experimentar entre diferentes compuestos usando el archivo "EOS_2_mezclas.ipynb" en donde se explica con más detalle las ecuaciones de estado, utilizando el siguiente libro: Reid, R. C., Prausnitz, J. M., & Poling, B. E. (1987). The properties of gases & liquids (4 th edition). McGraw-Hill.